### PR TITLE
Feature/apiv3/time entries

### PR DESCRIPTION
**Pull request by fupduck**
based upon https://github.com/opf/openproject/pull/6828
tries to reflect the requested changed by @ulferts 
I added a DELETE for time entries.

I did not find an `delete_time_entry` permission therefore used `edit_time_entry` - is this the correct way to go?

I hope this is somewhat complete and helps - I wrote ruby the last time 9 years ago…

Any comments appreciated!

Many thanks @oliverguenther for taking time for our request.

References issues:
https://community.openproject.com/projects/openproject/work_packages/29003/activity
https://community.openproject.com/projects/openproject/work_packages/29029/activity

### CODE REVIEW

**Comment by fupduck** [comment]
sorry for all the commits - plz squash when merging 😅

**Comment by oliverguenther** [comment]
Hi @fupduck , thanks for the effort. `edit_time_entries` is correct. You can see the mapping of permissions to controller actions in `config/initializers/permissions.rb`, which includes `destroy` for that permission.

File: app/policies/work_package_policy.rb [hunk 1]
```diff
@@ -120,7 +120,7 @@ def manage_subtasks_allowed?(work_package)
   def comment_allowed?(work_package)
     @comment_cache ||= Hash.new do |hash, project|
       hash[project] = user.allowed_to?(:add_work_package_notes, work_package.project) ||
-                      edit_allowed?(work_package)
+                      delete_allowed?(work_package)
```
Comment (ulferts) [inline app/policies/work_package_policy.rb:123]: As this is the work_package_policy, changing it in the scope of this PR looks quite fishy. I suspected this happened on a bulk search&replace.

File: .gitignore [hunk 2]
```diff
@@ -34,13 +34,17 @@
 
 # Ignore bundler config.
 /.bundle
+/vendor/bundle
```
Comment (ulferts) [inline .gitignore:37]: Why this?

File: app/services/time_entries/delete_service.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    self.time_entry = time_entry
+    self.contract_class = TimeEntries::DeleteContract
+  end
+
+  ##
+  # Deletes the given time entry if allowed.
+  #
+  # @return True if the deletion has been initiated, false otherwise.
+  def call
+    time_entry.destroy
```
Comment (ulferts) [inline app/services/time_entries/delete_service.rb:48]:
```suggestion
    validate_and_yield(time_entry, user) do
      time_entry.destroy
    end  
```

Without the `validate_and_yield` the contract is not actually used.

File: app/services/time_entries/update_service.rb [hunk 4]
```diff
@@ -29,48 +29,24 @@
...
 
   def set_attributes(attributes)
     time_entry.attributes = attributes
-
-    ##
-    # Update project context if moving time entry
-    if time_entry.work_package_id_changed?
-      time_entry.project_id = time_entry.work_package.project_id
-    end
-  end
```
Comment (ulferts) [inline app/services/time_entries/update_service.rb:56]: This needs to stay

**Review by ulferts — CHANGES_REQUESTED** [review]
First of all, @fupduck, thanks for building that end point, it is very much appreciated.

I skimmed over the PR and made notes where I see problems right away but didn't go into too many details as the PR you are building on by @oliverguenther has been rebased and I also issued another PR against Oliver's.

So once those base branches are consolidated, you should rebase on that. I'll post a message once that has happened.

**Comment by fupduck** [comment]
Thanks for the headsup.
Let me know when I can continue.

**Comment by ulferts** [comment]
@fupduck #6828 is in the dev branch now.

File: app/policies/work_package_policy.rb (hunk 1 above)
Comment (fupduck) [reply app/policies/work_package_policy.rb:123]: thanks

**Comment by fupduck** [comment]
Thanks for your support and your comments @ulferts and @oliverguenther 
Find the squashed and rebased MR here: https://github.com/opf/openproject/pull/6843
I hope I have adressed all open issues.

**Pull request closed without merging**